content/blog/symbolic-differentiation.Rmd

```diff
@@ -0,0 +1,396 @@
+  } else {
+    fn <- as.character(expr[[1]])
+    switch(fn,
+           "+" = d_plus(expr, name),
+           stop("not yet implemented"))
+  }
+}
+```
+
+
+With this we can differentiate a sum of any depth:
+
+```{r}
+differentiate(quote(x + 5), "x")
+differentiate(quote(x + y + x), "x")
+```
+
+We can then proceed, writing out rules for different functions as we need them. For example, the product rule:
+
+```{r}
+d_product <- function(expr, name) {
+  a <- expr[[2]]
+  b <- expr[[3]]
+  da <- differentiate(a, name)
+  db <- differentiate(b, name)
+  call("+", call("*", da, b), call("*", a, db))
+}
+```
+
+or the quotient rule
+
+```{r}
+d_quotient <- function(expr, name) {
+  a <- expr[[2]]
+  b <- expr[[3]]
+  da <- differentiate(a, name)
+  db <- differentiate(b, name)
+  ## da / b + a * db / (b * b)
+  call("-", call("/", da, b), call("/", call("*", a, db), call("*", b, b)))
+}
+```
+
+For subtraction, we need to distinguish between unary minus (e.g., `-a`) and subtraction (e.g., `a - b`)
+
+```{r}
+d_minus <- function(expr, name) {
+  if (length(expr) == 2) {
+    call("-", differentiate(expr[[2]], name))
+  } else {
+    call("-", differentiate(expr[[2]], name), differentiate(expr[[3]], name))
+  }
+}
+```
+
+It turns out that `(` is a function too, and also needs a rule, but it is very simple:
+
+```{r}
+d_parenthesis <- function(expr, name) {
+  call("(", differentiate(expr[[2]], name))
+}
+```
+
+We can put all these rules into a list:
+
+```{r}
+rules <- list(
+  "+" = d_plus,
+  "-" = d_minus,
+  "*" = d_product,
+  "/" = d_quotient,
+  "(" = d_parenthesis)
+```
+
+and rewrite our `differentiate()` implementation again:
+
+```{r}
+differentiate <- function(expr, name) {
+  if (!is.recursive(expr)) {
+    if (identical(expr, as.symbol(name))) 1 else 0
+  } else {
+    fn <- as.character(expr[[1]])
+    if (!(fn %in% names(rules))) {
+      stop(sprintf("Differentiation of '%s' not yet implemented", fn))
+    }
+    rules[[fn]](expr, name)
+  }
+}
+```
+
+and with this we can differentiate all sorts of things:
+
+```{r}
+differentiate(quote(-2 * x / (x * x - 3 * x)), "x")
+```
+
+This is fine, except that the generated expressions are fairly ugly, with lots of obviously redundant expressions (e.g., `-0 * x` which is obviously `0` and `+ -2 * 1` which is just `- 2`). However, the expressions agree with those from `D` once evaluated:
+
+```{r}
+eval(differentiate(quote(-2 * x / (x * x - 3 * x)), "x"), list(x = pi))
+eval(D(quote(-2 * x / (x * x - 3 * x)), "x"), list(x = pi))
+```
+
+## Extending the implementation by adding more rules
+
+We could extend this easily now by adding more rules, and the implementation will even tell us what we need to add. So if we try and evaluate
+
+```{r, error = TRUE}
+differentiate(quote(exp(2 * x)), "x")
+```
+
+we get told to we need to implement the rule for `exp` which is simply:
+
+```{r}
+d_exp <- function(expr, name) {
+  call("*", differentiate(expr[[2]], name), expr)
```

Comment:
So this rule means that the `expr` which is the second parameter to the `*` call isn't going to be differentiated any further so, that's going to make things a little different for expressions which contain other variables, which get differentiated away by the edge for all other rules (I think). So would they fail `eval`?

Reply:
This is me being slightly lazy because we could write

```
  call("*", differentiate(expr[[2]], name), call("exp", expr[[2]]))
```

which is how differentiation of things involving `exp()` looks, i.e., $d/dx \exp(f(x)) => f'(x) \exp(f(x))$ (see https://www.wolframalpha.com/input?i=differentiate+exp%28f%28x%29%29+wrt+x)